isisd/isis_northbound.c

```diff
@@ -1513,6 +1522,262 @@ static int isis_instance_mpls_te_router_address_destroy(enum nb_event event,
+isis_instance_segment_routing_msd_node_msd_modify(enum nb_event event,
+						  const struct lyd_node *dnode,
+						  union nb_resource *resource)
+{
+	struct isis_area *area;
+
+	if (event != NB_EV_APPLY)
+		return NB_OK;
+
+	area = nb_running_get_entry(dnode, NULL, true);
+	area->srdb.config.msd = yang_dnode_get_uint8(dnode, NULL);
+	isis_sr_cfg_msd_update(area);
+
+	return NB_OK;
+}
+
+static int
+isis_instance_segment_routing_msd_node_msd_destroy(enum nb_event event,
+						   const struct lyd_node *dnode)
+{
+	struct isis_area *area;
+
+	if (event != NB_EV_APPLY)
+		return NB_OK;
+
+	area = nb_running_get_entry(dnode, NULL, true);
+	area->srdb.config.msd = 0;
+	isis_sr_cfg_msd_update(area);
+
+	return NB_OK;
+}
+
+/*
+ * XPath: /frr-isisd:isis/instance/segment-routing/prefix-sid-map/prefix-sid
+ */
+static int isis_instance_segment_routing_prefix_sid_map_prefix_sid_create(
+	enum nb_event event, const struct lyd_node *dnode,
+	union nb_resource *resource)
+{
+	struct isis_area *area;
+	struct prefix prefix;
+	struct sr_prefix_cfg *pcfg;
+
+	if (event != NB_EV_APPLY)
+		return NB_OK;
+
+	area = nb_running_get_entry(dnode, NULL, true);
+	yang_dnode_get_prefix(&prefix, dnode, "./prefix");
+
+	pcfg = isis_sr_cfg_prefix_add(area, &prefix);
+	nb_running_set_entry(dnode, pcfg);
+
+	return NB_OK;
+}
+
+static int isis_instance_segment_routing_prefix_sid_map_prefix_sid_destroy(
+	enum nb_event event, const struct lyd_node *dnode)
+{
+	struct sr_prefix_cfg *pcfg;
+	struct isis_area *area;
+
+	if (event != NB_EV_APPLY)
+		return NB_OK;
+
+	pcfg = nb_running_unset_entry(dnode);
+	area = pcfg->area;
+	isis_sr_cfg_prefix_del(pcfg);
+	lsp_regenerate_schedule(area, area->is_type, 0);
+
+	return NB_OK;
+}
+
+static void
+isis_instance_segment_routing_prefix_sid_map_prefix_sid_apply_finish(
+	const struct lyd_node *dnode)
+{
+	struct sr_prefix_cfg *pcfg;
+	struct isis_area *area;
+
+	pcfg = nb_running_get_entry(dnode, NULL, true);
+	area = pcfg->area;
+	lsp_regenerate_schedule(area, area->is_type, 0);
+}
+
+/*
+ * XPath:
+ * /frr-isisd:isis/instance/segment-routing/prefix-sid-map/prefix-sid/sid-value-type
+ */
+static int
+isis_instance_segment_routing_prefix_sid_map_prefix_sid_sid_value_type_modify(
+	enum nb_event event, const struct lyd_node *dnode,
+	union nb_resource *resource)
+{
+	struct sr_prefix_cfg *pcfg;
+
+	if (event != NB_EV_APPLY)
+		return NB_OK;
+
+	pcfg = nb_running_get_entry(dnode, NULL, true);
+	pcfg->sid_type = yang_dnode_get_enum(dnode, NULL);
+
+	return NB_OK;
+}
+
+/*
+ * XPath:
+ * /frr-isisd:isis/instance/segment-routing/prefix-sid-map/prefix-sid/sid-value
+ */
+static int
+isis_instance_segment_routing_prefix_sid_map_prefix_sid_sid_value_modify(
+	enum nb_event event, const struct lyd_node *dnode,
+	union nb_resource *resource)
+{
+	struct sr_prefix_cfg *pcfg;
+
+	if (event != NB_EV_APPLY)
+		return NB_OK;
+
+	pcfg = nb_running_get_entry(dnode, NULL, true);
+	pcfg->sid = yang_dnode_get_uint32(dnode, NULL);
+
+	return NB_OK;
+}
```

Comment:
Yes, the local Prefix-SIDs aren't validated at configuration time (similar to IOS-XR). I think logging SID overflows to the terminal is a reasonable thing to do, but I'm not sure if the command should be rejected in that case. I need to think more about this.

Reply:
If we reject the command, would we not advertise and prefix sids at all? I suspect this is the correct answer, because -- we don't know which prefixes are "more important," and we want to make certain the user "sees" this error, and the behavior is consistent. Better for things to be totally broken than partially, I think, in this case.

One thing I might be missing here -- is it possible to use tags or some such to determine which prefixes have sids assigned? Is this worth doing?

Reply:
> If we reject the command, would we not advertise and prefix sids at all? I suspect this is the correct answer, because -- we don't know which prefixes are "more important," and we want to make certain the user "sees" this error, and the behavior is consistent. Better for things to be totally broken than partially, I think, in this case.

After my latest updates, isisd doesn't accept the configuration of invalid Prefix-SIDs anymore.

> One thing I might be missing here -- is it possible to use tags or some such to determine which prefixes have sids assigned? Is this worth doing?

There's a new `show isis segment-routing prefix-sids` command that should fit that purpose.

Example:
```
debian# show isis segment-routing prefix-sids 
Area 1:
 IS-IS L1 Prefix-SIDs:

 Prefix                SID  In Label  Nexthop                             Out Label  Uptime    
 ----------------------------------------------------------------------------------------------
 1.1.1.1/32            10   -         local                               -          -         
 2.2.2.2/32            20   16020     10.0.1.2, eth-sw1                   16020      00:00:50  
 3.3.3.3/32            30   16030     10.0.1.3, eth-sw1                   17030      00:00:50  
 4.4.4.4/32            40   16040     10.0.1.2, eth-sw1                   16040      00:00:50  
 5.5.5.5/32            50   16050     10.0.1.3, eth-sw1                   17050      00:00:50  
 6.6.6.6/32            60   16060     10.0.1.3, eth-sw1                   17060      00:00:50  
                                      10.0.1.2, eth-sw1                   16060      00:00:50  
 2001:db8:1000::1/128  11   -         local                               -          -         
 2001:db8:1000::2/128  21   16021     fe80::687b:b3ff:fe01:8df0, eth-sw1  16021      00:00:50  
 2001:db8:1000::3/128  31   16031     fe80::54d1:6aff:fed0:ded8, eth-sw1  17031      00:00:50  
 2001:db8:1000::4/128  41   16041     fe80::687b:b3ff:fe01:8df0, eth-sw1  16041      00:00:50  
 2001:db8:1000::5/128  51   16051     fe80::54d1:6aff:fed0:ded8, eth-sw1  17051      00:00:50  
 2001:db8:1000::6/128  61   16061     fe80::54d1:6aff:fed0:ded8, eth-sw1  17061      00:00:50  
                                      fe80::687b:b3ff:fe01:8df0, eth-sw1  16061      00:00:50
```